### PR TITLE
ActiveRecordInterface improvements

### PR DESCRIPTION
**Pull request by SamMousa**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes/no
| Related issues  | #9745

Small improvements to `ActiveRecordInterface` to make it more complete.
The following functions have been added. For each I have listed 1 "proof", one instance where it is called based on the assumption that the class is an active record class.

- populateRecord [ActiveQueryTrait#134](https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQueryTrait.php#L134)
- instantiate[ActiveQueryTrait#134](https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQueryTrait.php#L132)
- afterFind [ActiveQuery#223](https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQuery.php#L223)

### CODE REVIEW

**Comment by SamMousa** [comment]
The failing tests seem to not be originating in any code but in the test environment.

File: framework/db/ActiveRecordInterface.php [hunk 1]
```diff
@@ -427,4 +427,44 @@ public function unlink($name, $model, $delete = false);
...
+     */
+    public static function populateRecord($record, $row);
+
+    /**
+     * This method is called when the AR object is created and populated with the query result.
+     * The default implementation will trigger an [[EVENT_AFTER_FIND]] event.
+     * When overriding this method, make sure you call the parent implementation to ensure the
+     * event is triggered.
+     */
+    public function afterFind();
```
Comment (samdark) [inline framework/db/ActiveRecordInterface.php]: Doesn't look like it belongs here. Even if it does, there are other alike methods such as `afterSave()`.

**Review by samdark — CHANGES_REQUESTED** [review]

**Comment by cebe** [comment]
Thank you for the proposal.

> populateRecord ActiveQueryTrait#134
> instantiateActiveQueryTrait#134

while they are called within the trait, adding them to the interface it would limit the possibilities of implementing ARInterface in different ways that use different ways of instantiating records.
I do not think there is a good practical reason to add these to the interface, even if it would be theoretically more correct solution.

> afterFind ActiveQuery#223

This is the implementation bound to a concrete relalisation of ActiveRecordInterface for DB, its fine to expect a db\ActiveRecord here.

File: framework/db/ActiveRecordInterface.php (hunk 1 above)
Comment (SamMousa) [inline framework/db/ActiveRecordInterface.php]: @samdark, I've listed my "proofs" in the post. This is the exempt is from `ActiveQuery` line 221:
````
        if (!$this->asArray) {
            foreach ($models as $model) {
                $model->afterFind();
            }
````
If you call it without checking if it exists you're assuming it is there. Which means it should be in the interface.

Comment (klimov-paul) [inline framework/db/ActiveRecordInterface.php]: ActiveRecordTrait has been design for the `BaseActiveRecord` in order to reuse the code for common inplementations. It does not sticts to `ActiveRecordInterface` anyhow.

**Comment by SamMousa** [comment]
I feel the querying should be separate from the models. If the query related classes do something it should be part of the AR interface otherwise I cannot implement my own AR models that work with the query part.

File: framework/db/ActiveRecordInterface.php (hunk 1 above)
Comment (SamMousa) [inline framework/db/ActiveRecordInterface.php]: I'm not talking about `ActiveRecordTrait`?

**Comment by klimov-paul** [comment]
> I feel the querying should be separate from the models. If the query related classes do something it should be part of the AR interface otherwise I cannot implement my own AR models that work with the query part.

There are `ActiveRecordInterface` and `ActiveQueryInterface` thier method lists are consistent. If you create your own implementation you need to satisfy those interfaces.
`ActiveQueryTrait` is a particular implementation, to be precise - a shared part of particular implementation, which is bound to `BaseActiveRecord` class, which is a shared part of particular implementation.
If you wish to use `ActiveQueryTrait` you should either extend your `ActiveRecord` from `BaseActiveRecord` or implement it compatible way.
if you wish to create pure ActiveRecord/ActiveQuery implementation you should use nothing but `ActiveRecordInterface` and `ActiveQueryInterface`.

**Comment by SamMousa** [comment]
They are not consistent separately. If I implement `ActiveRecordInterface` in my model class. And try to retrieve instances of it via Yii's `ActiveQuery` concrete implementation, it will break because the `ActiveRecordInterface` is incomplete.

**Comment by klimov-paul** [comment]
> They are not consistent separately. If I implement ActiveRecordInterface in my model class. And try to retrieve instances of it via Yii's ActiveQuery concrete implementation, it will break because the ActiveRecordInterface is incomplete.

???
What are you actually trying to do? Are you attempt to use `yii\db\ActiveQuery` for `yii\mongodb\ActiveRecord` or something?

Particular ActiveRecord and ActiveQuery is **always** a pair. If while creating such a pair you can reuse code from `BaseActiveRecord` or `ActiveQueryTrait` it is fine, but if it does not fit - you should avoid them and craete implementation from scratch.

File: framework/db/ActiveRecordInterface.php (hunk 1 above)
Comment (klimov-paul) [inline framework/db/ActiveRecordInterface.php]: Here are **your** links from the PR description:

- populateRecord [ActiveQueryTrait#134](https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQueryTrait.php#L134)
- instantiate[ActiveQueryTrait#134](https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQueryTrait.php#L132)
- afterFind [ActiveQuery#223](https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQuery.php#L223)

2 are pointing to `ActiveQureyTrait` and 1 to `ActiveQuery` (the one specific for `yii\db\ActiveRecord` and **not shared** for `yii\mongodb\ActiveRecord` for instance).

**Comment by SamMousa** [comment]
> Particular ActiveRecord and ActiveQuery is always a pair. If while creating such a pair you can reuse code from BaseActiveRecord or ActiveQueryTrait it is fine, but if it does not fit - you should avoid them and craete implementation from scratch.

Why? There is no argument for this, just the fact that your interface is incomplete forces them to be a pair...
If the interface is complete, then by definition concrete implementations are exchangeable.

**Comment by klimov-paul** [comment]
> Why? There is no argument for this, just the fact that your interface is incomplete forces them to be a pair...
If the interface is complete, then by definition concrete implementations are exchangeable.

Very well, lets start from beginning then: can you please open a famous book by Martin Fowler named " Patterns of Enterprise Application Architecture" and draw a class diagram for 'ActiveRecord' pattern? How does it look like?

**Comment by SamMousa** [comment]
That's a non-argument. You've split the query logic away from the `ActiveRecord` class. There is no argument to then couple the interface when there's no need to.

Anyway, I don't feel like there is a point to try and change your mind, it hasn't ever happened before. Instead I'll try to convince the team members who take the time to think for themselves instead of blindly quoting Martin Fowler like he wrote the bible.

**Comment by klimov-paul** [comment]
> That's a non-argument.

I can say just the same for yours: as I have already said [here](https://github.com/yiisoft/yii2/pull/14295#issuecomment-309365717) at the present state you are talking that `yii\mongodb\ActiveRecord` should be actually able to use `yii\db\ActiveQuery` for the models search. This can not happen anyhow

> You've split the query logic away from the ActiveRecord class. There is no argument to then couple the interface when there's no need to.

This is a common practice and used in other places too. For example, see:
https://github.com/yiisoft/yii2/blob/master/framework/db/mssql/TableSchema.php#L22
https://github.com/yiisoft/yii2/blob/master/framework/db/mssql/Schema.php#L160

Here we can also tell, that `$catalogName` should belong to `\yii\db\TableSchema` to produce consistent `TableSchema` class interface, even if it not used anywhere besides MSSQL.
Here we can say your argument: why I can not take plain `yii/db/TableSchema` an use it with `yii/db/mssql/Schema`?
Should we change this and other places as well to make you feel better?

> Anyway, I don't feel like there is a point to try and change your mind, it hasn't ever happened before. Instead I'll try to convince the team members who take the time to think for themselves instead of blindly quoting Martin Fowler like he wrote the bible.

It is a pity you do not want to re-think the issue from beginning with me.
The point is an ActiveRecord is a monolith: it incapsulates BOTH finding and saving the data. It can not function otherwise - that is why it is called ACTIVE record actually. Creating a separated ActiveQuery was a design choice, which simplifies and structures the implementation. And even still ActiveRecord actually hides ActiveQuery behind the scenes: developer uses `ActiveRecord::find()` to instatiate an ActiveQuery instead of `new ActiveQuery()`.

Here you wish to extend `ActiveQueryInterface` with extra methods, which puts a restriction on the possible particular implementation. You enforce `afterFind` method, which according to its specification should trigger a Yii event, while particular ActiveRecord library may be built on non-Yii classes an use its own event managenemt or do not use events at all. You are converting a necessity into a requirement/restriction here.

Your argument is there should not be any `such method does not exist` error during using any combination of ActiveRecord and ActiveQuery, while random combnation of thoses will definetely produce other fatal errors.

**Comment by cebe** [comment]
@SamMousa I agree with @klimov-paul in the point that there is not reason in separating ActiveRecord and ActiveQuery implementations. AR and AQ are a set of classes that play together, i.e. highly depend on each other. There is a balance to take between perfect interface completeness (formal correctness) and practical usage. Yii is mostly following the practical path. Unless you have a concrete implementation that would show up a problem with the current interface I would prefer to not make any changes to the interface.

**Comment by samdark** [comment]
I agree with @cebe and @klimov-paul. In this particular case initial proof seem to be incorrect.

**Pull request closed without merging**